### DISCUSSION TITLE
Relays on dingtian 4ch switch on off on start or reboot with Tasmota

### DISCUSSION THREAD

**Discussion by jeroenst** [arendst/Tasmota#21266] [Q&A] [answered]
I have a dingtian 4ch board. When I start or reboot the board it switches the relays on for a few milliseconds. 

According to the manufacturer, the gpio connected to the relays are provided with 4k7 pull down resistors. It seems that Tasmota initializes the output high instead of low causing the relays to switch on and off.

Can I change the default gpio state on boot?

![-1088284157-349877494](https://github.com/arendst/Tasmota/assets/2833940/428edc45-31c0-4bbc-aa5f-186c3aa4fa8e)

Program Version	13.4.0(TasmoCompiler-esp32generic)
Build Date & Time	2024-04-18T20:44:36
Core/SDK Version	2_0_14/4.4.6.240105

**Comment by barbudor** [comment]
What template are you using ? Relay or Relay_Inverted ?
What is your PowerOnState for each power ?
Can you try to draw the exact schematic from a GPIO to the relay ?

To insure you recover the power state after a Restart, you should have `Setoption1 1`
To insure you recover the saved power state after a power cycle, you should have `PowerOnState 3`

**Comment by jeroenst** [comment]
The manufacturer send me their original firmware which doesn't turn on the relays on boot and reboot so it's not the hardware that causes this. 

I tried both, relay and relay inverted, makes no difference except the inverted output state. 
PowerOnState 1 or 3 makes not difference.
Setoption1 1 or setoption1 0 Makes no difference.

I also see that the only relay that goes on for a few milliseconds at boot is connected to gpio16. The other relays stay off at boot as they should.

I didn't get a schematic yet from the manufacturer, I asked for it again.

**Reply by Tabaluga63** [reply]
Hello jeroenst, 
can you share the original firmware for the Dingtian 4chrelay here?
I have the board too and flashed with Tasmota, it works fine, but i will compare some functions with original firmware. I don´t find the original firmware in www.

**Reply by jeroenst** [reply]
I don't have the original firmware.

**Comment by jeroenst** [comment]
Here you can see the behavior on reboot : 

https://github.com/arendst/Tasmota/assets/2833940/d1e04e1c-d8e7-4a8a-bbba-ac9d3e104565

**Comment by Noschvie** [comment]
Same as #21056

**Reply by jeroenst** [reply]
So it is an Tasmota issue, and the software of the manufacturer doesn't have this issue, so it's possible to overcome this problem.

Is there a define to overcome this problem and disable PSRAM on GPIO16?

**Reply by barbudor** [reply]
So your problem is only on GPIO16?
You originally were not so specific

For the diagram, a papier, a pencil, a multimeter and your eyes should be enough. I don't expect the manufacturer to share such details.

**Reply by jeroenst** [reply]
True, before I thought I heard all relays but when I removed the cover I saw it was only 1.

**Comment by Jason2866** [comment]
If it is "only" GPIO 16 it is an expected behaviour. Tasmota firmware is a general build and does check at start for PSRAM (which is connect to the GPIOs 16&17). This generates blinking at start for this two GPIOs.
You can use the Solo1 development firmware (based on core 3.0.0) which has no support for PSRAM.
Not working with release branch (still using core 2.0.14). There is support for PSRAM still enabled!

**Reply by jeroenst** [reply]
Is it possible to put the psram check between #ifndef DISABLE_PSRAMCHECK so it can be disabled on build?

**Reply by Jason2866** [reply]
You can try PR #21282

**Comment by Jason2866** [comment]
@jeroenst Please dont discuss individual issues in a PR.  Close VSC and delete hidden folder `.platformio`. Clone my repo (https://github.com/Jason2866/Tasmota.git) and checkout path `disable_psram_check`.
Start VSC and "Open Folder" with the fresh clone.
The error you see is a defect Platformio installation

**Comment by Jason2866** [comment, marked as answer]
@jeroenst Compiled from the branch
[tasmota32-nopsram.zip](https://github.com/arendst/Tasmota/files/15114189/tasmota32-nopsram.zip)

**Reply by jeroenst** [reply]
Thank you! This works, the relay keeps quiet while (re)booting.

Now I have to find out why I get a build error on your branch.

**Reply by jeroenst** [reply]
A clean clone of your branch and removing .platformio did the trick and the building now succeeds. Thank you.

**Reply by une18** [reply]
> Thank you! This works, the relay keeps quiet while (re)booting.


You got it solved. What am I doing wrong?

**Comment by Noschvie** [comment]
Hello,
Is the ETH interface supported by Tasmota or how can it be put into operation?
Thanks!

**Reply by Jason2866** [reply]
Use this template https://templates.blakadder.com/dingtian_DT-R004.html

**Comment by Noschvie** [comment]
> ![-1088284157-349877494](https://private-user-images.githubusercontent.com/2833940/325222557-428edc45-31c0-4bbc-aa5f-186c3aa4fa8e.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxMzAwODEsIm5iZiI6MTczMjEyOTc4MSwicGF0aCI6Ii8yODMzOTQwLzMyNTIyMjU1Ny00MjhlZGM0NS0zMWMwLTRiYmMtYWE1Zi0xODZjM2FhNGZhOGUuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjBUMTkwOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTExODk0ZjgzNGEwYTI4MzA4N2JlMGQ1Yzc5ZTdkZGJmNDU2M2NlNjQxZTkxMTVhNDc4NDU0YjNhMzAzZDJlYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.enZMSITQ8BwRTdnkcHsReQxljLOsVkTWjW6pKyceYQ4)
> 

@jeroenst  What does "NRST" mean? Thanks

**Reply by jeroenst** [reply]
Not Reset (inverted reset line, low = reset, high = run)

**Reply by une18** [reply]
I am confused with this board. Curent template in Blackaddre does not have ETH. I am able to get ETH working, but can not get rid of the relay clicking....

I have flashed the NoPsram file above, but my Relay 1 keeps klicking on boot

**Reply by jeroenst** [reply]
Can you maybe share your template with Ethernet so it can be changed in the documentation?

**Reply by jeroenst** [reply]
Is relay 1 on our off before reboot? I can imagine it will click when it's on.

**Reply by une18** [reply]
> Can you maybe share your template with Ethernet so it can be changed in the documentation?

Got hints in Discord from @Jason2866 

Use the template https://templates.blakadder.com/dingtian_DT-R004.html
and add for Ethernet 
GPIO 0 Output HI
GPIO18 ETH MDIO
GPIO23 ETH MDC

change or check 
EthAddress 0
EthClockMode 3
EthType 7

**Reply by une18** [reply]
> Is relay 1 on our off before reboot? I can imagine it will click when it's on.

I turn relays ON, to see if they click during reboot. And Relay 1 on Gpio16 does

**Reply by jeroenst** [reply]
And when all the Relais are off do they still click?

**Reply by une18** [reply]
> And when all the Relais are off do they still click

No. Only Relay1 is clicking when in ON state. If Relay1 is OFF it does not click. Relay 2 never does.

**Reply by jeroenst** [reply]
I don't know if this is hardware or software related. In the io table of the esp32 there are no remarks on GPIO 16 according to state on boot. Maybe @Jason2866 can help with this?

**Reply by Jason2866** [reply]
@arendst Could it be that a GPIO PSRAM reset is made? Should not happen / be done since PSRAM is disabled and the GPIOs should not have "special" treatment in that case.

**Reply by arendst** [reply]
Wasn't this https://github.com/arendst/Tasmota/discussions/21266#discussioncomment-9228017 the solution?

**Reply by une18** [reply]
For some reason does not work for me with the same file from April. Another board will arrive in coming days, can test if it acts the same. Current was ordered 13 months ago. It is V2 not V1

**Reply by Jason2866** [reply]
@arendst Yes, but since that time we have the changed function (code) to re-init the PSRAM GPIOs

**Reply by arendst** [reply]
But it is skipped for solo1:
```
#ifndef CORE32SOLO1
  // restore GPIO5/18 or 16/17 if no PSRAM is found which may be used by Ethernet among others
  if (!FoundPSRAM()) {
    // test if the CPU is not pico
    uint32_t pkg_version = bootloader_common_get_chip_ver_pkg();
    if (pkg_version <= 3) {         // D0WD, S0WD, D2WD
      gpio_reset_pin((gpio_num_t)CONFIG_D0WD_PSRAM_CS_IO);
      gpio_reset_pin((gpio_num_t)CONFIG_D0WD_PSRAM_CLK_IO);
      // IDF5.3 fix esp_gpio_reserve used in init PSRAM
      esp_gpio_revoke(BIT64(CONFIG_D0WD_PSRAM_CS_IO) | BIT64(CONFIG_D0WD_PSRAM_CLK_IO));
    }
  }
#endif  // CORE32SOLO1
```
so a solo firmware should just work.

**Reply by Jason2866** [reply]
Yes, testing with the solo1 build is for sure a good idea. Nevertheless we should think about (if this is the reason) how to solve for standard esp32 build.

EDIT: adding additional the guard `# ifndef DISABLE_PSRAMCHECK` should work?

**Reply by arendst** [reply]
Currently testing with a dev board. Default ESP32 code does toggle relay on GPIO16. Testing solo1....

**Reply by arendst** [reply]
Yep. SIngle core doesn't toggle the relay. So the OP should use latest solo binary on his board.

**Reply by une18** [reply]
OK! Will try to compile later and test it out.

**Reply by Jason2866** [reply]
No need to compile. Just use the solo1 build.

**Reply by arendst** [reply]
> EDIT: adding additional the guard # ifndef DISABLE_PSRAMCHECK should work?

@Jason2866 testing...

**Reply by arendst** [reply]
@Jason2866 yes. Works too. So I'll change it like this
```
#ifndef FIRMWARE_SAFEBOOT
#ifndef DISABLE_PSRAMCHECK
#ifndef CORE32SOLO1
  // restore GPIO5/18 or 16/17 if no PSRAM is found which may be used by Ethernet among others
  if (!FoundPSRAM()) {
    // test if the CPU is not pico
    uint32_t pkg_version = bootloader_common_get_chip_ver_pkg();
    if (pkg_version <= 3) {         // D0WD, S0WD, D2WD
      gpio_reset_pin((gpio_num_t)CONFIG_D0WD_PSRAM_CS_IO);
      gpio_reset_pin((gpio_num_t)CONFIG_D0WD_PSRAM_CLK_IO);
      // IDF5.3 fix esp_gpio_reserve used in init PSRAM
      esp_gpio_revoke(BIT64(CONFIG_D0WD_PSRAM_CS_IO) | BIT64(CONFIG_D0WD_PSRAM_CLK_IO));
    }
  }
#endif  // CORE32SOLO1
#endif  // DISABLE_PSRAMCHECK
#endif  // FIRMWARE_SAFEBOOT
```

**Reply by goodkiller** [reply]
> @Jason2866 yes. Works too. So I'll change it like this
> 
> ```
> #ifndef FIRMWARE_SAFEBOOT
> #ifndef DISABLE_PSRAMCHECK
> #ifndef CORE32SOLO1
>   // restore GPIO5/18 or 16/17 if no PSRAM is found which may be used by Ethernet among others
>   if (!FoundPSRAM()) {
>     // test if the CPU is not pico
>     uint32_t pkg_version = bootloader_common_get_chip_ver_pkg();
>     if (pkg_version <= 3) {         // D0WD, S0WD, D2WD
>       gpio_reset_pin((gpio_num_t)CONFIG_D0WD_PSRAM_CS_IO);
>       gpio_reset_pin((gpio_num_t)CONFIG_D0WD_PSRAM_CLK_IO);
>       // IDF5.3 fix esp_gpio_reserve used in init PSRAM
>       esp_gpio_revoke(BIT64(CONFIG_D0WD_PSRAM_CS_IO) | BIT64(CONFIG_D0WD_PSRAM_CLK_IO));
>     }
>   }
> #endif  // CORE32SOLO1
> #endif  // DISABLE_PSRAMCHECK
> #endif  // FIRMWARE_SAFEBOOT
> ```

It worked! Just be sure that gets in into newer Tasmota version

**Reply by une18** [reply]
With Solo and some tweaking the relays are silent now!  Thank You @arendst @Jason2866 @jeroenst !

**Reply by jeroenst** [reply]
I got this error while building TASMOTA32 with DISABLE_PSRAMCHECK:
```
/home/jeroen/git/Tasmota/tasmota/tasmota_support/support_esp32.ino:588:111: error: operator '||' has no right operand
  588 | #if CONFIG_IDF_TARGET_ESP32C2 || CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32C6 || DISABLE_PSRAMCHECK || CORE32SOLO1
```

Also while building the SOLO1 variant.

**Reply by jeroenst** [reply]
I had to do "#define DISABLE_PSRAMCHECK true" instead

**Reply by une18** [reply]
@jeroenst We have been working slowly on the device. Most recent template for 2ch boards: 
 {"NAME":"DTWonder_2relay","GPIO":[3840,9408,226,9440,6209,0,0,0,224,0,0,0,225,0,5600,0,0,0,0,5568,0,0,0,0,0,0,0,0,0,9952,35,0,160,0,0,161],"FLAG":0,"BASE":1}
Defined additionally:
Gpio34 as Button4 to be able to reset the device. (Factory button on board)
Gpio32 as LED to be able to make it signal error codes

In our case Relay1 is RS485, Relay2 is physical Relay1 and Relay3 is physical Relay2. 

4ch and 8ch boards are on the way.

To avoid problems we automatically turn OFF Wifi when LAN is  connected.
